File: 1Lab/SinglyList.cpp
#include "SinglyList.h"
#include <iostream>

using namespace std;

void FPRINT(const SinglyLinkedList& list) {
    FNode* nd = list.head; 
    while (nd != nullptr) {
        cout << "\"" << nd -> key << "\" ";
        nd = nd->next;
    }
    cout << endl;
}

void FPRINT_REC(FNode* nd) { 
    if (nd == nullptr) {
        cout << endl;
        return;
    }
    cout << "\"" << nd -> key << "\" ";
    FPRINT_REC(nd -> next);
}

void FPUSH_HEAD(SinglyLinkedList& list, const string& key) {
    FNode* newNode = new FNode{key, list.head}; 
    list.head = newNode;
    list.size++;
}

void FPUSH(SinglyLinkedList& list, const string& key) {
    FNode* newNode = new FNode{key, nullptr}; 
    if (!list.head) {
        list.head = newNode;
        list.size++;
        return;
    }
    FNode* temp = list.head; 
    while (temp -> next != nullptr)
        temp = temp -> next;
    temp -> next = newNode;
    list.size++;
}

void FPUSH_AFTER(FNode* nd, const string& key) {
    if (!nd) return;
    FNode* newNode = new FNode{key, nd -> next}; 
    nd -> next = newNode;
}

void FPUSH_BEFORE(SinglyLinkedList& list, const string& target, const string& key) {
    if (!list.head) return;
    if (list.head -> key == target) {
        FPUSH_HEAD(list, key);
        return;
    }
    FNode* prev = list.head; 
    FNode* curr = list.head -> next; 
    while (curr && curr -> key != target) {
        prev = curr;
        curr = curr -> next;
    }
    if (!curr) return;
    FNode* newNode = new FNode{key, curr}; 
    prev -> next = newNode;
    list.size++;
}

void FDEL_HEAD(SinglyLinkedList& list) {
    if (!list.head) return;
    FNode* temp = list.head; 
    list.head = list.head -> next;
    delete temp;
    list.size--;
}

void FDEL_TAIL(SinglyLinkedList& list) {
    if (!list.head) return;
    if (!list.head -> next) {
        delete list.head;
        list.head = nullptr;
        list.size--;
        return;
    }
    FNode* temp = list.head; 
    while (temp -> next -> next)
        temp = temp -> next;
    delete temp->next;
    temp -> next = nullptr;
    list.size--;
}

void FDEL_AFTER(FNode* nd) {
    if (!nd || !nd -> next) return;
    FNode* del = nd -> next; 
    nd -> next = del -> next;
    delete del;
}

void FDEL_BEFORE(SinglyLinkedList& list, const string& target) {
    if (!list.head || list.head -> key == target) return;
    FNode* prevPrev = nullptr; 
    FNode* prev = list.head; 
    FNode* curr = list.head->next; 
    while (curr && curr -> key != target) {
        prevPrev = prev;
        prev = curr;
        curr = curr -> next;
    }
    if (!curr) return; 
    if (!prevPrev) { 
        FDEL_HEAD(list);
    } else {
        prevPrev -> next = curr;
        delete prev;
        list.size--;
    }
}

void FDEL(SinglyLinkedList& list, const string& value) {
    if (!list.head) return;
    if (list.head -> key == value) {
        FDEL_HEAD(list);
        return;
    }
    FNode* prev = list.head; 
    FNode* curr = list.head->next; 
    while (curr && curr -> key != value) {
        prev = curr;
        curr = curr -> next;
    }
    if (!curr) return;
    prev -> next = curr->next;
    delete curr;
    list.size--;
}

FNode* FGET(const SinglyLinkedList& list, const string& value) {
    FNode* temp = list.head;
    while (temp) {
        if (temp -> key == value)
            return temp;
        temp = temp -> next;
    }
    return nullptr;
}

File: 1LabGo/singly_list.go
package main

import "fmt"

type SinglyNode struct {
	key  string
	next *SinglyNode
}

type SinglyList struct {
	head *SinglyNode
	size int
}

func (sl *SinglyList) FPushHead(key string) {
	newNode := &SinglyNode{key: key, next: sl.head}
	sl.head = newNode
	sl.size++
}

func (sl *SinglyList) FPushTail(key string) {
	newNode := &SinglyNode{key: key}
	if sl.head == nil {
		sl.head = newNode
	} else {
		current := sl.head
		for current.next != nil {
			current = current.next
		}
		current.next = newNode
	}
	sl.size++
}

func (sl *SinglyList) FGet(key string) *SinglyNode {
	current := sl.head
	for current != nil {
		if current.key == key {
			return current
		}
		current = current.next
	}
	return nil
}

func (sl *SinglyList) FPushAfter(targetKey, newKey string) bool {
	targetNode := sl.FGet(targetKey)
	if targetNode == nil {
		return false
	}
	newNode := &SinglyNode{key: newKey, next: targetNode.next}
	targetNode.next = newNode
	sl.size++
	return true
}

func (sl *SinglyList) FPushBefore(targetKey, newKey string) bool {
	if sl.head == nil {
		return false
	}
	// Если целевой элемент - голова
	if sl.head.key == targetKey {
		sl.FPushHead(newKey)
		return true
	}
	// Ищем узел перед целевым
	prev := sl.head
	for prev.next != nil && prev.next.key != targetKey {
		prev = prev.next
	}
	if prev.next == nil {
		return false
	}
	newNode := &SinglyNode{key: newKey, next: prev.next}
	prev.next = newNode
	sl.size++
	return true
}

func (sl *SinglyList) FDelHead() (string, error) {
	if sl.head == nil {
		return "", fmt.Errorf("ошибка: список пуст")
	}
	value := sl.head.key
	sl.head = sl.head.next
	sl.size--
	return value, nil
}

func (sl *SinglyList) FDelTail() (string, error) {
	if sl.head == nil {
		return "", fmt.Errorf("ошибка: список пуст")
	}
	if sl.head.next == nil {
		value := sl.head.key
		sl.head = nil
		sl.size--
		return value, nil
	}
	current := sl.head
	for current.next.next != nil {
		current = current.next
	}
	value := current.next.key
	current.next = nil
	sl.size--
	return value, nil
}

func (sl *SinglyList) FDel(key string) bool {
	if sl.head == nil {
		return false
	}
	if sl.head.key == key {
		sl.head = sl.head.next
		sl.size--
		return true
	}
	prev := sl.head
	for prev.next != nil && prev.next.key != key {
		prev = prev.next
	}
	if prev.next != nil {
		prev.next = prev.next.next
		sl.size--
		return true
	}
	return false
}

func (sl *SinglyList) FPrint() {
	if sl.head == nil {
		fmt.Println("Список пуст.")
		return
	}
	current := sl.head
	for current != nil {
		fmt.Printf(`"%s" `, current.key)
		current = current.next
	}
	fmt.Println()
}
